src/node.rs

```diff
@@ -118,7 +120,69 @@ impl Node {
                      compute_current_candidate_blocks(potentially_current));
-        self.current_blocks = compute_current_blocks(&self.current_candidate_blocks);
+
+        mem::replace(&mut self.current_blocks,
+                     compute_current_blocks(&self.current_candidate_blocks))
+
+    }
+
+    fn merge_messages(&self, prev_current_blocks: &CurrentBlocks)
+        -> Vec<Message> {
+        let mut messages = vec![];
+        let new_current_blocks = &self.current_blocks - prev_current_blocks;
+
+        for our_new_block in section_blocks(&new_current_blocks, self.our_name) {
+            for our_prev_block in section_blocks(&prev_current_blocks, self.our_name) {
+                // Merge case.
+                if let Some(sibling_pfx) = our_prev_block.prefix.sibling() {
+                    if our_new_block.prefix == our_prev_block.prefix.popped() {
+                        // In the case of a merge, send our sibling's history to all the
+                        // sections we are connected to but they are not.
+                        // FIXME: need to send history for all descendents of sibling prefix?
+                        // i.e. if we are 0 and want to merge to 1, but other
+                        // sections are 11, 100, 101.
+                        let sibling_block = prev_current_blocks.iter()
+                            .find(|block| block.prefix == sibling_pfx)
+                            .expect(
+                                &format!("oh gee: {:#?}", prev_current_blocks)
+                            );
+
+                        let disconn_from_sibling = prev_current_blocks.iter()
+                            .filter(|block| !block.prefix.is_neighbour(&sibling_pfx))
+                            .flat_map(|block| block.members.iter().map(|n| *n));
+
+                        let segment = chain_segment(sibling_block, &self.vote_counts);
+
+                        let new_messages = disconn_from_sibling.map(|neighbour| {
+                            Message {
+                                sender: self.our_name,
+                                recipient: neighbour,
+                                content: VoteBundle(segment.clone()),
+                            }
+                        });
+                        messages.extend(new_messages);
+                    }
+                }
+            }
+        }
+
+        messages
+    }
+
+    /// Drop blocks for sections that we aren't neighbours of.
+    fn prune_split_blocks(&mut self) {
+        let all_current_blocks = mem::replace(&mut self.current_blocks, btreeset!{});
+
+        let our_prefix = section_blocks(&all_current_blocks, self.our_name)
+            .next()
+            .unwrap()
+            .prefix;
+
+        for block in all_current_blocks {
+            if block.prefix.is_neighbour(&our_prefix) || block.prefix == our_prefix {
+                self.current_blocks.insert(block);
+            }
+        }
```

Comment:
Yeah I'm fairly sure it's ok, the block still exists in the sets of valid blocks, and candidate current blocks.